src/Control/Effect/Carrier.hs

```diff
@@ -52,3 +63,31 @@ handlePure handler = hmap handler . fmap' handler
 handleCoercible :: (HFunctor sig, Coercible f g) => sig f (f a) -> sig g (g a)
 handleCoercible = handlePure coerce
 {-# INLINE handleCoercible #-}
+
+{-# DEPRECATED handleReader, handleState, handleEither, handleTraversable
+  "Compose carrier types from other carriers and define 'eff' with handleCoercible instead" #-}
+
+-- | Thread a @Reader@-like carrier through an 'HFunctor'.
+handleReader :: HFunctor sig => r -> (forall x . f x -> r -> g x) -> sig f (f a) -> sig g (g a)
+handleReader r run = handlePure (flip run r)
+{-# INLINE handleReader #-}
+
+-- | Thread a @State@-like carrier through an 'Effect'.
+handleState :: Effect sig => s -> (forall x . f x -> s -> g (s, x)) -> sig f (f a) -> sig g (g (s, a))
+handleState s run = handle (s, ()) (uncurry (flip run))
+{-# INLINE handleState #-}
+
+-- | Thread a carrier producing 'Either's through an 'Effect'.
+handleEither :: (Carrier sig g, Effect sig) => (forall x . f x -> g (Either e x)) -> sig f (f a) -> sig g (g (Either e a))
+handleEither run = handle (Right ()) (either (pure . Left) run)
+{-# INLINE handleEither #-}
+
+-- | Thread a carrier producing values in a 'Traversable' 'Monad' (e.g. '[]') through an 'Effect'.
+handleTraversable :: (Effect sig, Applicative g, Monad m, Traversable m) => (forall x . f x -> g (m x)) -> sig f (f a) -> sig g (g (m a))
+handleTraversable run = handle (pure ()) (fmap join . traverse run)
+{-# INLINE handleTraversable #-}
+
+-- | A backwards-compatibility shim, equivalent to 'id'.
+interpret :: carrier a -> carrier a
+interpret = id
+{-# DEPRECATED interpret "Not necessary with monadic carriers; remove or replace with 'id'." #-}
```

Comment:
Love the thought you’ve put into definitions like these and the deprecation messages 👍